src/libdredd/src/mutation_replace_unary_operator.cc

```diff
@@ -0,0 +1,263 @@
+  std::string new_function_name = GetFunctionName(ast_context);
+  std::string result_type = unary_operator_.getType()
+                                ->getAs<clang::BuiltinType>()
+                                ->getName(ast_context.getPrintingPolicy())
+                                .str();
+  std::string input_type = unary_operator_.getSubExpr()
+                               ->getType()
+                               ->getAs<clang::BuiltinType>()
+                               ->getName(ast_context.getPrintingPolicy())
+                               .str();
+
+  // To avoid problems of ambiguous function calls, the argument types (ignoring
+  // whether they are references or not) are baked into the mutation function
+  // name. Some type names have space in them (e.g. 'unsigned int'); such spaces
+  // are replaced with underscores.
+
+  ApplyTypeModifiers(unary_operator_.getSubExpr(), input_type);
+  ApplyTypeModifiers(&unary_operator_, result_type);
+
+  clang::SourceRange unary_operator_source_range_in_main_file =
+      GetSourceRangeInMainFile(preprocessor, unary_operator_);
+  assert(unary_operator_source_range_in_main_file.isValid() &&
+         "Invalid source range.");
+  clang::SourceRange input_source_range_in_main_file =
+      GetSourceRangeInMainFile(preprocessor, *unary_operator_.getSubExpr());
+  assert(input_source_range_in_main_file.isValid() && "Invalid source range.");
+
+  // Replace the unary operator expression with a call to the wrapper
+  // function.
+  //
+  // Subtracting |first_mutation_id_in_file| turns the global mutation id,
+  // |mutation_id|, into a file-local mutation id.
+  const int local_mutation_id = mutation_id - first_mutation_id_in_file;
+  bool result = rewriter.ReplaceText(
+      unary_operator_source_range_in_main_file,
+      new_function_name + "([&]() -> " + input_type + " { return " +
+          // We don't need to static cast constant expressions
+          (unary_operator_.getSubExpr()->isCXX11ConstantExpr(ast_context)
+               ? rewriter.getRewrittenText(input_source_range_in_main_file)
+               : ("static_cast<" + input_type + ">(" +
+                  rewriter.getRewrittenText(input_source_range_in_main_file) +
+                  ")")) +
+          "; }" + ", " + std::to_string(local_mutation_id) + ")");
+  (void)result;  // Keep release-mode compilers happy.
+  assert(!result && "Rewrite failed.\n");
+
+  std::vector<clang::UnaryOperatorKind> operators = {
+      clang::UnaryOperatorKind::UO_PreInc, clang::UnaryOperatorKind::UO_PostInc,
+      clang::UnaryOperatorKind::UO_PreDec, clang::UnaryOperatorKind::UO_PostDec,
+      clang::UnaryOperatorKind::UO_Not,    clang::UnaryOperatorKind::UO_Minus,
+      clang::UnaryOperatorKind::UO_LNot};
+
+  std::string new_function;
+  new_function = GenerateMutatorFunction(new_function_name, result_type,
+                                         input_type, operators, mutation_id);
```

Comment:
Given that this function is only ever called with one set of operators, it would be better to get rid of the `operators` parameter and bake this set of operators into `GenerateMutatorFunction`.

Reply:
Moved.